p2p/protocol/identify/obsaddr.go

```diff
@@ -127,43 +139,76 @@ func (oas *ObservedAddrSet) Addrs() (addrs []ma.Multiaddr) {
 	return addrs
 }
 
-// Add attemps to queue a new observed address to be added to the set.
-func (oas *ObservedAddrSet) Add(observed, local, observer ma.Multiaddr,
-	direction network.Direction) {
+// Record records an address observation, if valid.
+func (oas *ObservedAddrManager) Record(conn network.Conn, observed ma.Multiaddr) {
 	select {
-	case oas.wch <- newObservation{observed: observed, local: local, observer: observer, direction: direction}:
+	case oas.wch <- newObservation{
+		conn:     conn,
+		observed: observed,
+	}:
 	default:
-		log.Debugf("dropping address observation of %s; buffer full", observed)
+		log.Debugw("dropping address observation due to full buffer",
+			"from", conn.RemoteMultiaddr(),
+			"observed", observed,
+		)
 	}
 }
 
-func (oas *ObservedAddrSet) worker(ctx context.Context) {
+func (oas *ObservedAddrManager) teardown() {
+	oas.host.Network().StopNotify((*obsAddrNotifiee)(oas))
+
+	oas.mu.Lock()
+	oas.refreshTimer.Stop()
+	oas.mu.Unlock()
+}
+
+func (oas *ObservedAddrManager) worker() {
+	defer oas.teardown()
+
 	ticker := time.NewTicker(GCInterval)
 	defer ticker.Stop()
 
+	closingCh := oas.host.Network().Process().Closing()
```

Comment:
this feels more awkward than taking in a context to close on

Reply:
We need to pass the host in anyways... but I can just use both, I guess.